### DISCUSSION TITLE
Redux toolkit with getStaticProps and preloadedState how

### DISCUSSION THREAD

**Discussion by TommySorensen** [vercel/next.js#11192] [Help] [answered]
Does anyone have a good example of Next used with `getStaticProps` and Redux hooks together with a `preloadedState`?
This example does not use `getStaticProps` https://github.com/zeit/next.js/tree/canary/examples/with-redux

I have tried with this. It does set the preloaded state, but the client have a different state, so i end up 2 different states. I must be missing something.
```js
// store.js

export default function configureStore(preloadedState) {
  return configureStore({
    reducer,
    middleware,
    preloadedState,
  })
}
```

```js
// pages/_app.js

import store from './../store'
class MyApp extends App {
  render() {
    const { Component, pageProps } = this.props

    return (
      <ReduxProvider store={configureStore(pageProps)}>
          <Component {...pageProps} />
      </ReduxProvider>
    )
  }
}
```

This is also related to https://github.com/zeit/next.js/pull/11103 just with redux toolkit

**Comment by thomasdtucker** [comment]
Does this have a solution yet? I am seeing getInitialProps being used even in the redux examples in the next.js main repo as well or they are using next-redux-wrapper which specifically states that it breaks all Auto Partial Static Exports.

**Reply by TommySorensen** [reply]
Nope its not. And apparently we should not do it like i suggested accordingly to https://github.com/zeit/next.js/pull/11103#issuecomment-601775316 
That solution is similar to this https://github.com/zeit/swr/blob/master/examples/server-render/pages/%5Bpokemon%5D.js using useSWR hook. But i would like to get it working with redux toolkit.
This is the how the toolkit ended up being implemented: https://github.com/zeit/next.js/tree/canary/examples/with-redux-toolkit

**Reply by Butonix** [reply]
This example also doesn't support getStaticProps

**Comment by okeybond** [comment]
This would be helpful, please assist nextjs

**Comment by puneetv05** [comment]
any updates on this?  @TommySorensen did you solved this issue?

**Reply by TommySorensen** [reply]
My solution was to not use preloadedState in the build. It does not make sense to put it in redux at all. Just get the props from the page and be happy with that 🙂
This https://github.com/vercel/next.js/pull/11103#issuecomment-601775316 comment kind of make me realize, do i even need to put the all that data in redux? I think not. I have just released a website that just used props for the data, and that worked fine.

**Reply by puneetv05** [reply]
In my case  I can follow your approach check for the user is logged in and pass the prop to the state for further hydration to show logged in header or guest header. [#11033(comment)](https://github.com/vercel/next.js/pull/11103#issuecomment-601775316) make total sense.

**Reply by TommySorensen** [reply]
Yes but you will never check for the user state in `getStaticProps` function. You will do on hydrate instead. Basically look at the experience on https://vercel.com/dashboard when you are logget in. They load the user data on hydrate and mounts the data from several other endpoints. Maybe take a look at this example where they use iron-session https://github.com/vercel/next.js/blob/canary/examples/with-iron-session/pages/profile-sg.js, i found that example a very nice 😉

**Reply by puneetv05** [reply]
thanks @TommySorensen I'm thinking to implement via redux toolkit + [next-redux-wrapper](https://github.com/kirill-konshin/next-redux-wrapper) my site is not static btw vercel experience is legit! 
for Static props I found this: [next-redux-wrapper with redux toolkit example](https://github.com/kirill-konshin/next-redux-wrapper/issues/162#issuecomment-651157282) I think in this example we can use `getStaticProps` in place of `getServerSideProps` (haven't tried myself)

**Reply by TommySorensen** [reply]
Please give `getStaticProps` a try. You might feel it's the bomb 😉
I am working on a site where almost everything but some cms content is on pr. user basic. And i still feel the static thing is the way to do it. Does this not fit your need?:

````
// https://nextjs.org/docs/basic-features/data-fetching
When should I use getStaticProps?
You should use getStaticProps if:

The data required to render the page is available at build time ahead of a user’s request.
The data comes from a headless CMS.
The data can be publicly cached (not user-specific).
The page must be pre-rendered (for SEO) and be very fast — getStaticProps generates HTML and JSON files, both of which can be cached by a CDN for performance.

**Reply by puneetv05** [reply]
Would love to utilize cache for SEO will give `getStaticProps` a try ✌ thanks again!

**Reply by puneetv05** [reply]
given `getStaticProps` a try works great 😃 with Redux toolkit used [next-redux-wrapper](https://github.com/kirill-konshin/next-redux-wrapper) (with-redux-toolkit-app)

**Reply by TommySorensen** [reply]
Yes but the next-redux-wrapper is just a big load of empty state to do what? 😉

**Reply by puneetv05** [reply]
I just want to use redux-toolkit to work with SSR 😃

**Reply by puneetv05** [reply]
btw @TommySorensen is next-redux-wrapper not a good fit for getStaticProps with Redux toolkit?

**Reply by TommySorensen** [reply]
Yes it is a good fit, and you can make it work. But how will you share state between pages when the pages are generated on build? You will end up with a state from the last build since its static, just like some of the state examples in the next examples on github

**Reply by puneetv05** [reply]
yha but can we not hydrate the state ? https://github.com/vercel/next.js/pull/11103 looking into the comments it seems like we don't have a approach to hyderate the state via staticProps

**Reply by puneetv05** [reply]
FYI: for the example repo with redux-toolkit I have added & it works
```
//  lib/slices/clockSlice.js
 extraReducers: {
    [HYDRATE]: (state, action) => action.payload.clock,
  },
```

```
// pages/index.js
export const getStaticProps = wrapper.getStaticProps(async ({ store }) => {
  store.dispatch(tick({ light: false, lastUpdate: Date.now() }));
});
```

**Reply by TommySorensen** [reply]
I am not totally sure since i never had it working and i moved away from it. I only use redux from a clean state that is not preloaded. This was enough for me. Something like this, and then just default redux toolkit stuff.
```js
// ./pages/__app.js
import store from './store/store'

const App = ({ Component, pageProps }) => (
  <Provider store={store}>
      <Component {...pageProps} />
  </Provider>
)
```

```js
// ./store/store.js
import { configureStore } from '@reduxjs/toolkit'

import somethingReducer from './slices/something'

export default configureStore({
  reducer: {
    cool: somethingReducer,
  },
  devTools: true,
})

```

```js
// ./store/slices/something.js
import { createSlice } from '@reduxjs/toolkit'

const initialState = {
  something: null
}

const testSlice = createSlice({
  name: 'something',
  initialState,
  reducers: {
    setSomething: (state, action) => {
      state.something = action.payload
    }
  }
})

// Selectors
export const selectSomething = state => state.something

// Actions
export const { setSomething } = testSlice.actions

export default testSlice.reducer

```

**Reply by puneetv05** [reply]
> FYI: for the example repo with redux-toolkit I have added & it works
> 
> ```
> //  lib/slices/clockSlice.js
>  extraReducers: {
>     [HYDRATE]: (state, action) => action.payload.clock,
>   },
> ```
> 
> ```
> // pages/index.js
> export const getStaticProps = wrapper.getStaticProps(async ({ store }) => {
>   store.dispatch(tick({ light: false, lastUpdate: Date.now() }));
> });
> ```

you can try this approach IMO

**Reply by TommySorensen** [reply]
> FYI: for the example repo with redux-toolkit I have added & it works
> 
> ```
> //  lib/slices/clockSlice.js
>  extraReducers: {
>     [HYDRATE]: (state, action) => action.payload.clock,
>   },
> ```
> 
> ```
> // pages/index.js
> export const getStaticProps = wrapper.getStaticProps(async ({ store }) => {
>   store.dispatch(tick({ light: false, lastUpdate: Date.now() }));
> });
> ```

If that works, then i think you should create a example and do a PR for next so other can see the solution?

**Reply by puneetv05** [reply]
full example https://github.com/kirill-konshin/next-redux-wrapper/issues/162#issuecomment-651157282 can work with `getServerSideProps` or `getStaticProps` as per need 😉

**Reply by puneetv05** [reply]
@TommySorensen will definitely do a PR with good use case example, suggestion appreciated 😃

**Reply by TommySorensen** [reply]
@puneetverma05 Try to use it with the `getStaticPaths` and see if it still works as expected. I never got that part to work. I ended up having shared state between pages or something like that. Can't remember

**Reply by TommySorensen** [reply]
> @TommySorensen will definitely do a PR with good use case example, suggestion appreciated 😃

Did you find a good way to handle the `SERVER_ACTION`and `CLIENT_ACTION`? I mean when you browse on client side, then you will need to merge the state with the state from the build.

**Reply by puneetv05** [reply]
@TommySorensen nope haven't got the time to play with getStaticPaths 😢

**Comment by julien-amblard** [comment, marked as answer]
Hi there,
seems like they updated their doc 
https://github.com/vercel/next.js/blob/canary/examples/with-redux/pages/ssg.js

you can now return a `initialReduxState` object in props

```js
import Page from '../components/page'

export default function SSG() {
  return <Page />
}

export function getStaticProps() {
  return {
    props: {
      initialReduxState: {
        lastUpdate: Date.now(),
        light: false,
      },
    },
  }
}
```
it work for me

**Reply by TommySorensen** [reply]
Interesting. Seems like they went with merging the state from build time to whatever state you have from your previous visited page. I guess that could work 🧐.
https://github.com/vercel/next.js/blob/canary/examples/with-redux/store.js#L52:L61
Thanks for the heads up 💪

**Reply by julien-amblard** [reply]
It work well, BUT, I've find out it create a new devtools instance everytime you came throught this (same for `getServerSideProps`), looks like a weird behaviour to me, but not sure if it's because of this way to use it, or if it's a devtools bug. 
I've find nothing about this issue anywhere, may be I'm the only one who think it's weird (or may be I don't know how to google)

¯\\\_(ツ)_/¯

If someone have an idea, I'll take it
